GitHub 로그인 / 회원 가입 기능 구현

kimyubi opened the pull request:
#### 구현한 내용 : https://github.com/DONMIT/DONMIT_API/issues/1, https://github.com/DONMIT/DONMIT_API/issues/2

#### 구현한 기능 테스트:

1) 아직 클라이언트가 구현되지 않았기 때문에 테스트를 위해 GitHub Oauth app에서 redirect_url을 
    http://localhost:8080/login/oauth2/code/github로 설정하고 서버에서 임시 코드를 받아 진행하였습니다.
![image](https://user-images.githubusercontent.com/62657545/222650209-511e4950-9499-4ec2-acc9-50d339f7f2d6.png)


2) 따라서, http://[리다이렉트 URL]?code=f34625fc66333c09865d 와 같이 URL을 입력하면 서버에서 비즈니스 로직을 수행하고 우리 서비스의 액세스 토큰과 리프레쉬 토큰을 반환합니다.
![image](https://user-images.githubusercontent.com/62657545/222650906-84dfd535-eb86-44d1-b7ef-c2ee1cf5eb5c.png)

3) GitHub OAuth + Spring Security + JWT 를 사용하여 인증을 구현한 후  TestController를 만들어 인증이 정상적으로
    동작하는지 테스트하였습니다.
![image](https://user-images.githubusercontent.com/62657545/222651531-80d60017-6641-4794-91c6-0c37952261d5.png)

4) 본인의 GitHub 계정으로 우리 서비스에 로그인 했을 때 DB에 GitHub 사용자 정보가 저장되는 모습입니다.
![image](https://user-images.githubusercontent.com/62657545/222651677-3c74ec2a-1783-4d79-b9bb-ba328c92021b.png)


#### To Do
- 리팩토링
- Spring Rest Docs 도입

---------------
#### 논의하고 싶은 점
아래 사진은 GitHub 사용자 정보를 받아오기 위한 DTO의 필드입니다.

GitHub 사용자 정보는 json 형식으로 반환 되기 때문에 json 데이터를 Java 객체로 변환하는 과정에서 
avatar_url와 같이 스네이크 케이스를 불가피 사용하게 되는데 이러한 점을 피하고자 필드의 이름은 카멜 케이스로
하되,  @JsonProperty("avatar_url")과 같은 방법을 사용하였습니다.

하지만 위와 같은 방법을 사용해도  **CheckStyle** 경고가 발생합니다.  

![image](https://user-images.githubusercontent.com/62657545/222652198-f49dde8b-3cb0-428a-ac6b-ded6751a6684.png)
![image](https://user-images.githubusercontent.com/62657545/222652921-35912a44-bd63-4fbb-bee6-ede76bac4196.png)

Commit:
Feat #1: Base Entity 작성
- 생성 일자, 수정 일자를 기록하기 위한 BaseEntity 클래스 작성
- - JPA의 Auditing 기능을 이용하기 위해 JpaAudtingConfig 설정 파일 작성

Commit:
Feat #1: DefaultRes 작성
- 공통 response DTO인 DefaultRes 클래스 작성

Commit:
Feat #2: GitHub 로그인 / 회원가입 기능 구현

Commit:
Refactor: DTO에 Record 클래스 적용
- e.printStackTrace() 대체
- setter 사용 지양

Commit:
Refactor #2: JWT 인증 과정에서 발생하는 예외 처리

Commit:
Refactor : Record 적용한 DTO에 requireNonNull 사용

Commit:
.gitignore 파일에 /.idea/ 추가

Commit:
rename : .gitignore 파일에 /.idea/ 추가

Commit:
Refactor: Fixed untracked files

Commit:
Refactor: Fixed untracked files

Commit:
Style : 디렉토리 구조 변경 (계층형 -> 도메인형)

Commit:
Merge remote-tracking branch 'origin/feature/GitHub_Oauth' into feature/GitHub_Oauth

Commit:
Refactor #4: WebClientConfig 작성

- WebClientConfig에서 gitHubWebClient을 Bean으로 등록하여 GitHub API를 호출할 때 사용한다.

- in-memory buffer 값, 로깅, 타임 아웃 설정을 위해 Builder 클래스를 통해 WebClient를 생성한다.

Commit:
Refactor #4: AuthService의 getAccessToken 메소드 리팩토링

- RestTemplate를 통해 GitHub API를 활용하던 방식에서,
  WebClient를 통해 GitHub API를 사용하는 방식으로 전환

Commit:
Refactor #4: AuthService의 findProfile 메소드 리팩토링

- RestTemplate를 통해 GitHub API를 활용하던 방식에서,
  WebClient를 통해 GitHub API를 사용하는 방식으로 전환

Commit:
Refactor #: 인증이 필요하지 않은 URI는 JWT 관련 필터를 타지 않도록 리팩토링

0n1dev commented on src/main/java/dev/donmit/donmitapi/auth/api/AuthApi.java:
```diff
@@ -0,0 +1,43 @@
+			return new ResponseEntity<>(res, HttpStatus.OK);
+		}
+
+		return new ResponseEntity<>(
+			DefaultResponse.response(HttpStatus.OK.value(), "로그인 실패 ( => GitHub Access Token 발급 실패)"),
+			HttpStatus.OK);
```
커스텀 응답 객체와 ResponseEntity를 같이 사용하다 보면 제네릭이 중첩으로 사용되어 코드의 가독성이 떨어질 것 같은데 더 좋은 방법이 없을까요?

kimyubi replied:
커스텀 응답 객체만 사용하는 것이 더 좋을 것 같습니다!

0n1dev commented on src/main/java/dev/donmit/donmitapi/auth/application/AuthService.java:
```diff
@@ -0,0 +1,81 @@
+	public OauthTokenResponseDto getAccessToken(String code, String clientId,
+		String clientSecret) {
+		MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
+		params.add("client_id", clientId);
+		params.add("client_secret", clientSecret);
+		params.add("code", code);
```
상수를 사용하는게 좀 더 좋아보입니다.

kimyubi replied:
확인했습니다.

0n1dev commented on src/main/java/dev/donmit/donmitapi/auth/application/AuthService.java:
```diff
@@ -0,0 +1,81 @@
+		params.add("code", code);
+
+		return gitHubWebClient.mutate()
+			.baseUrl("https://github.com")
+			.build()
+			.post()
+			.uri("/login/oauth/access_token")
```
baseUrl uri는 properties에 명시 해놓고 사용하는게 더 좋지 않을까 생각됩니다.

kimyubi replied:
확인했습니다.

0n1dev commented on src/main/java/dev/donmit/donmitapi/auth/application/AuthService.java:
```diff
@@ -0,0 +1,81 @@
+		TokenResponseDto tokenResponseDto = jwtTokenProvider.createToken(user.getGithubLogin(), user.getGithubId());
+		// Refresh Token 저장
+		user.saveRefreshToken(tokenResponseDto.refreshToken());
+
+		return DefaultResponse.response(HttpStatus.OK.value(), "로그인 성공",
```
응답 상태 코드와 그 상태 코드에 대한 메시지를 가지고 있는 Enum 클래스를 만들면 어떨까요?

kimyubi replied:
넵 확인했습니다!

0n1dev commented on src/main/java/dev/donmit/donmitapi/auth/dto/OauthTokenResponseDto.java:
```diff
@@ -0,0 +1,12 @@
+public record OauthTokenResponseDto(String access_token, String scope, String token_type) {
+
+	public OauthTokenResponseDto {
+		Objects.requireNonNull(access_token);
+		Objects.requireNonNull(scope);
+		Objects.requireNonNull(token_type);
```
위에서는 == null 로 체크를 하셨는데 아래에서는 requireNonNull를 통해서 체크하셨네요.
다르게 체크하신 이유가 있으신가요?

kimyubi replied:
위에서는 == null 로 체크를 했다고 말씀하신 부분이 어떤 부분인지  잘 모르겠습니다!

0n1dev commented on src/main/java/dev/donmit/donmitapi/auth/util/JwtTokenProvider.java:
```diff
@@ -0,0 +1,124 @@
+			.signWith(SignatureAlgorithm.HS256, secretKey)
+			.compact();
+
+		return TokenResponseDto.builder()
+			.grantType("bearer")
```
이것도 상수로 빼주는게 좋아 보이네요

kimyubi replied:
확인했습니다.

0n1dev commented on src/main/java/dev/donmit/donmitapi/global/common/response/DefaultResponse.java:
```diff
@@ -0,0 +1,37 @@
+import lombok.Builder;
+import lombok.Getter;
+
+@Getter
+@Builder
```
정적 팩토리 메서드로 따로 인스턴스 생성을 해주고 계신데 빌더를 따로 또 추가하신 이유가 있으신가요?

kimyubi replied:
DefaultResponse 클래스 내의 response 메소드에서 보다 가독성 좋게 DefaultResponse 인스턴스를 생성하고자  @Builder를 사용하였습니다. 

그러나, DefaultResponse 클래스 외부에서 Builder를 사용하여 인스턴스를 생성할 가능성이 있으니, 이 부분은 정적 팩토리 메소드를 사용해야만 인스턴스를 생성할 수 있도록 수정하도록 하겠습니다.

0n1dev commented on src/main/java/dev/donmit/donmitapi/global/config/security/CustomAuthenticationEntryPoint.java:
```diff
@@ -0,0 +1,40 @@
+		response.setContentType(MediaType.APPLICATION_JSON_VALUE);
+
+		// 응답 객체 초기화
+		body.put("status", HttpServletResponse.SC_UNAUTHORIZED);
+		body.put("error", "Unauthorized");
+		body.put("message", authException.getMessage());
+		body.put("path", request.getServletPath());
```
이 친구들도 상수로 가면 어떨까요?

kimyubi replied:
확인했습니다.

0n1dev commented on src/main/java/dev/donmit/donmitapi/global/config/webclient/WebClientConfig.java:
```diff
@@ -0,0 +1,48 @@
+
+@Configuration
+@Slf4j
+public class WebClientConfig {
+	@Bean
```
class 상단부는 개행 처리를 해주면 좀 더 가독성이 있습니다.

kimyubi replied:
참고하여 수정하겠습니다.

0n1dev commented on src/main/java/dev/donmit/donmitapi/global/config/webclient/WebClientConfig.java:
```diff
@@ -0,0 +1,48 @@
+	@Bean
+	public WebClient gitHubWebClient() {
+		// in-memory buffer 값 설정
+		ExchangeStrategies exchangeStrategies = ExchangeStrategies.builder()
+			.codecs(configurer -> configurer.defaultCodecs().maxInMemorySize(2 * 1024 * 1024))
```
상수 처리 해주시는게 좀 더 명확해 보입니다.

kimyubi replied:
확인했습니다.

Commit:
style, Refactor:

- 응답 상태 코드와 그 상태 코드에 대한 메시지를 가지고 있는 Enum 클래스인 CustomResponse 작성

- 상수를 정의하는 Constants 클래스 작성
- DefaultResponse 클래스에서 @Builder를 제거하여 클래스 외부에서 DefaultResponse 인스턴스를 생성하지 못하도록 리팩토링

Commit:
style

Commit:
style
